File: lang/ts/tests/mermaid-view.test.ts
import { describe, it, expect } from 'vitest';
import { Tramli } from '../src/tramli.js';
import { MermaidGenerator } from '../src/mermaid-generator.js';
import { flowKey } from '../src/flow-key.js';
import type { StateProcessor, StateConfig } from '../src/types.js';
import type { FlowContext } from '../src/flow-context.js';

// Issue #47: view option for unified state / dataflow entry point.

type S = 'a' | 'b' | 'c';
const KNumber = flowKey<number>('Num');
const KString = flowKey<string>('Str');

const stateConfig: Record<S, StateConfig> = {
  a: { terminal: false, initial: true },
  b: { terminal: false },
  c: { terminal: true },
};

const p1: StateProcessor<S> = {
  name: 'p1',
  requires: [],
  produces: [KNumber],
  process(ctx: FlowContext) { ctx.put(KNumber, 1); },
};

const p2: StateProcessor<S> = {
  name: 'p2',
  requires: [KNumber],
  produces: [KString],
  process(ctx: FlowContext) { ctx.put(KString, `n=${ctx.get(KNumber)}`); },
};

function buildFlow() {
  return Tramli.define<S>('view-test', stateConfig)
    .from('a').auto('b', p1)
    .from('b').auto('c', p2)
    .build();
}

describe('Issue #47: MermaidGenerator view option', () => {
  it('default view is state (stateDiagram-v2)', () => {
    const def = buildFlow();
    const out = MermaidGenerator.generate(def);
    expect(out).toContain('stateDiagram-v2');
    expect(out).toMatch(/a --> b/);
  });

  it('view: "state" is equivalent to default', () => {
    const def = buildFlow();
    expect(MermaidGenerator.generate(def, { view: 'state' })).toBe(
      MermaidGenerator.generate(def),
    );
  });

  it('view: "dataflow" produces flowchart with FlowKey edges', () => {
    const def = buildFlow();
    const out = MermaidGenerator.generate(def, { view: 'dataflow' });
    expect(out).toContain('flowchart LR');
    expect(out).toContain('p1');
    expect(out).toContain('Num');
    expect(out).toContain('produces');
    expect(out).toContain('requires');
    expect(out).not.toMatch(/stateDiagram/);
  });

  it('view: "dataflow" === generateDataFlow()', () => {
    const def = buildFlow();
    expect(MermaidGenerator.generate(def, { view: 'dataflow' })).toBe(
      MermaidGenerator.generateDataFlow(def),
    );
  });
});

File: lang/java/src/test/java/org/unlaxer/tramli/MermaidViewTest.java
package org.unlaxer.tramli;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Issue #47: unified view option for MermaidGenerator.
 */
final class MermaidViewTest {

    enum S implements FlowState {
        A(false, true),
        B(false, false),
        C(true, false);

        private final boolean terminal, initial;
        S(boolean terminal, boolean initial) { this.terminal = terminal; this.initial = initial; }
        @Override public boolean isTerminal() { return terminal; }
        @Override public boolean isInitial() { return initial; }
    }

    static final StateProcessor P1 = new StateProcessor() {
        @Override public String name() { return "p1"; }
        @Override public Set<Class<?>> requires() { return Set.of(); }
        @Override public Set<Class<?>> produces() { return Set.of(Integer.class); }
        @Override public void process(FlowContext ctx) { ctx.put(Integer.class, 1); }
    };

    static final StateProcessor P2 = new StateProcessor() {
        @Override public String name() { return "p2"; }
        @Override public Set<Class<?>> requires() { return Set.of(Integer.class); }
        @Override public Set<Class<?>> produces() { return Set.of(String.class); }
        @Override public void process(FlowContext ctx) { ctx.put(String.class, "n=" + ctx.get(Integer.class)); }
    };

    private static FlowDefinition<S> buildFlow() {
        return Tramli.<S>define("view-test", S.class)
                .from(S.A).auto(S.B, P1)
                .from(S.B).auto(S.C, P2)
                .build();
    }

    @Test void defaultViewIsState() {
        var def = buildFlow();
        String out = MermaidGenerator.generate(def);
        assertTrue(out.contains("stateDiagram-v2"));
        assertTrue(out.contains("A --> B"));
    }

    @Test void viewStateEqualsDefault() {
        var def = buildFlow();
        assertEquals(
                MermaidGenerator.generate(def),
                MermaidGenerator.generate(def, MermaidGenerator.View.STATE));
    }

    @Test void viewDataflowProducesFlowchart() {
        var def = buildFlow();
        String out = MermaidGenerator.generate(def, MermaidGenerator.View.DATAFLOW);
        assertTrue(out.contains("flowchart LR"), out);
        assertTrue(out.contains("p1"), out);
        assertTrue(out.contains("Integer"), out);
        assertTrue(out.contains("produces"), out);
        assertTrue(out.contains("requires"), out);
        assertFalse(out.contains("stateDiagram"), out);
    }

    @Test void viewDataflowEqualsGenerateDataFlow() {
        var def = buildFlow();
        assertEquals(
                MermaidGenerator.generateDataFlow(def),
                MermaidGenerator.generate(def, MermaidGenerator.View.DATAFLOW));
    }
}
